File: LeetCode/Hard/Binary Tree Maximum Path Sum/Go/bt_max_path_sum.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val, Left: nil, Right: nil} }

func get_max_path_sum(root *TreeNode) int {
	ans := math.MinInt64
	recur(root, &ans)
	return ans
}

func recur(root *TreeNode, ans *int) int {
	if root == nil {
		return 0
	}
	left := recur(root.Left, ans)
	right := recur(root.Right, ans)
	l_sum, r_sum := max(root.Val, root.Val+left), max(root.Val, root.Val+right)
	total_sum := l_sum + r_sum - root.Val
	*ans = max(*ans, total_sum)
	return max(l_sum, r_sum)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	root := New(1)
	root.Left = New(2)
	root.Right = New(3)
	fmt.Println(get_max_path_sum(root))
}

File: LeetCode/Hard/Binary Tree Maximum Path Sum/C/bt_max_path_sum.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

typedef struct TreeNode TreeNode;

int recur(TreeNode* root, int *res) {
    if (root == NULL) return 0;
    int left = recur(root->left, res);
    int right = recur(root->right, res);
    int l_sum = MAX(root->val, left + root->val);
    int r_sum = MAX(root->val, right + root->val);
    int total = l_sum + r_sum - root->val;
    *res = MAX(*res, total);
    return MAX(l_sum, r_sum);
}

int maxPathSum(TreeNode* root){
    int res = -(1e9+7);
    recur(root, &res);
    return res;
}
